x/ibc/06-solomachine/misbehaviour.go

```diff
@@ -0,0 +1,69 @@
+package solomachine
+
+import (
+	"bytes"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	clientexported "github.com/cosmos/cosmos-sdk/x/ibc/02-client/exported"
+	clienttypes "github.com/cosmos/cosmos-sdk/x/ibc/02-client/types"
+)
+
+// CheckMisbehaviourAndUpdateState determines whether or not the currently registered
+// public key signed over two different messages with the same sequence. If this is true
+// the client state is updated to a frozen status.
+func CheckMisbehaviourAndUpdateState(
+	clientState clientexported.ClientState,
+	_ clientexported.ConsensusState,
+	misbehaviour clientexported.Misbehaviour,
+) (clientexported.ClientState, error) {
+
+	// cast the interface to specific types before checking for misbehaviour
+	smClientState, ok := clientState.(ClientState)
+	if !ok {
+		return nil, sdkerrors.Wrapf(clienttypes.ErrInvalidClientType, "client state type %T is not solo machine", clientState)
+	}
+
+	if smClientState.IsFrozen() {
+		return nil, sdkerrors.Wrapf(clienttypes.ErrClientFrozen, "client is already frozen")
+	}
+
+	evidence, ok := misbehaviour.(Evidence)
+	if !ok {
+		return nil, sdkerrors.Wrapf(clienttypes.ErrInvalidClientType, "evidence type %T is not solo machine", misbehaviour)
+	}
+
+	if err := checkMisbehaviour(smClientState, evidence); err != nil {
+		return nil, err
+	}
+
+	smClientState.Frozen = true
+	return smClientState, nil
+}
+
+// checkMisbehaviour checks if the currently registered public key has signed
+// over two different messages at the same sequence.
+func checkMisbehaviour(clientState ClientState, evidence Evidence) error {
+	pubKey := clientState.ConsensusState.PubKey
+
+	// assert that provided signature data are different
+	if bytes.Equal(evidence.SignatureOne.Data, evidence.SignatureTwo.Data) {
+		return sdkerrors.Wrap(clienttypes.ErrInvalidEvidence, "evidence signatures have identical data messages")
+	}
+
+	data := append(sdk.Uint64ToBigEndian(evidence.Sequence), evidence.SignatureOne.Data...)
+
+	// check first signature
+	if !pubKey.VerifyBytes(data, evidence.SignatureOne.Signature) {
+		return sdkerrors.Wrap(clienttypes.ErrInvalidEvidence, "evidence signature one failed to be verified by currently registered public key")
+	}
```

Comment:
nit, move to a `checkSignature(sequence uint64, signature bytes) error` as described here: https://github.com/cosmos/ics/blob/master/spec/ics-006-solo-machine-client/README.md#misbehaviour-predicate

Reply:
Will add it to a `types/proof.go` file that also constructs sign bytes as per suggestion mentioned below